Extensions/v0.7 extensions (#86)
kwilteam/docs: docs/extensions/kuneiform-precompiles/example.mdx

---
sidebar_position: 2
sidebar_label: "Example: SHA hashing"
id: precompile-example
title: "Example: SHA Hashing Precompile"
description: Implementing a precompile for SHA hash functions
slug: /extensions/precompiles/sha
---

> The following example demonstrates how to implement a precompile for the SHA256 hash function.

Reviewer: ```suggestion
The following example demonstrates how to implement a precompile for the SHA256 and SHA512 hash functions.
```

type HashExtension struct {
	algorithm string
}

// Call executes the requested method of the precompile.
// It has one method, "hash", which takes a string as an argument,
// and returns the hex encoded hash of the string.
func (h *HashExtension) Call(scoper *actions.ProcedureContext, app *common.App, method string, inputs []any) ([]any, error) {
	// there is only one method, "hash"
	if method != "hash" {
		return nil, fmt.Errorf("method %s not found", method)
	}

	// there should only be one input, and it should be a string
	if len(inputs) != 1 {
		return nil, fmt.Errorf("hash: expected 1 string argument, got %d", len(inputs))
	}

	str, ok := inputs[0].(string)
	if !ok {
		return nil, fmt.Errorf("hash: expected string argument, got %T", inputs[0])
	}

	var hashBytes []byte
	// hash the string
	switch h.algorithm {
	case "sha256":
		b := sha256.Sum256([]byte(str))
		hashBytes = b[:]
	case "sha512":
		b := sha512.Sum512([]byte(str))
		hashBytes = b[:]
	default:
		// this should never happen
		return nil, fmt.Errorf("algorithm %s not found", h.algorithm)
	}

	// return the hash as a string
	return []any{hex.EncodeToString(hashBytes)}, nil
}
```

## Step 2: Use The Precompile

> The precompile can be used in a Kuneiform schema. The following example demonstrates how to use the precompile in a Kuneiform schema.

Reviewer: ```suggestion
The next step is to call the precompile in a Kuneiform schema. The following example demonstrates how to use the precompile in a Kuneiform schema.
```

> It will have two actions: "SHA256" and "SHA512". Each action will take a single input, which will be a string of the data to hash.

Reviewer: ```suggestion
The schema will have two actions: "SHA256" and "SHA512". Each action will take a single input, which will be a string of the data to hash.
```

> The action will return the hex encoded hash of the input data. Since no data is being stored, the actions will both be public and view.

Reviewer: ```suggestion
The action will return the hex encoded hash of the input data. Since the actions do not store data, they will be public and view.
```
